### PR TITLE
Yaksa integration

### PR DESCRIPTION
**Pull request by pavanbalaji**
## Pull Request Description

This PR integrates the yaksa library into mpich, behind the typerep interface.  It redirects all typerep calls to the yaksa API.  For the cases where yaksa doesn't provide direct support, the typerep glue layer adds the necessary functionality.

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [x] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

### CODE REVIEW

**Comment by pavanbalaji** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by shawnccx** [comment]
@pavanbalaji 
Do you need me to rebase https://github.com/pmodels/mpich/pull/4455 on this, and delete its temporary buffer solution? Though https://github.com/pmodels/mpich/pull/4455 is due today.

**Comment by pavanbalaji** [comment]
> @pavanbalaji
> Do you need me to rebase #4455 on this, and delete its temporary buffer solution? Though #4455 is due today.

That's up to you.  If #4455  is ready, you can merge it in.  But once this PR is integrated, you'd need to remove the direct calls to yaksa and use typerep.

**Comment by pavanbalaji** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by gcongiu** [comment]
I am seeing multiple errors with fortran types when compiling with `--disable-fortran`:
```
../src/mpi/datatype/typerep/src/typerep_yaksa_init.c:49:9: error: duplicate case value
         case MPI_INTEGER:
         ^
../src/mpi/datatype/typerep/src/typerep_yaksa_init.c:28:9: error: previously used here
         case MPI_CHARACTER:
         ^
```

**Comment by pavanbalaji** [comment]
> I am seeing multiple errors with fortran types when configuring with `--disable-fortran`:
> 
> ```
> ../src/mpi/datatype/typerep/src/typerep_yaksa_init.c:49:9: error: duplicate case value
>          case MPI_INTEGER:
>          ^
> ../src/mpi/datatype/typerep/src/typerep_yaksa_init.c:28:9: error: previously used here
>          case MPI_CHARACTER:
>          ^
> ```

Thanks.  I'll look into it.

**Comment by pavanbalaji** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by pavanbalaji** [comment]
I have no idea why the build failed.  I'm not able to reproduce the compile errors.  Maybe something got fixed when I rebased.  Rerunning the tests.

**Comment by pavanbalaji** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

File: src/mpi/datatype/typerep/src/typerep_yaksa_create.c [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+     * need to use yaksa subarray types */
+    return MPI_SUCCESS;
+}
+
+int MPIR_Typerep_create_darray(int size, int rank, int ndims, const int *array_of_gsizes,
+                               const int *array_of_distribs, const int *array_of_dargs,
+                               const int *array_of_psizes, int order, MPI_Datatype oldtype,
+                               void **typerep)
+{
+    return MPI_SUCCESS;
```
[the remark this answers is not in the document]
Comment (pavanbalaji) [reply src/mpi/datatype/typerep/src/typerep_yaksa_create.c:242]: `darray` is different: yaksa doesn't have a darray datatype.  But it does have a `subarray` datatype, which we are not using in MPICH.

**Comment by pavanbalaji** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by pavanbalaji** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by pavanbalaji** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by pavanbalaji** [comment]
Two sets of tests seem to be failing:

1. One set is because yaksa does not return how many bytes were actually unpacked in a `yaksa_iunpack` call.  MPICH needs this information to determine cases where there's a datatype mismatch.  This needs to be fixed in yaksa (see https://github.com/pmodels/yaksa/issues/67).

2. The second set is based on how `ch4` handles noncontiguous data in MPI RMA.  Both `ch4:ofi` and `ch4` AM layer seem to use inefficient paths for this.  Both layers convert noncontiguous data to IOVs, even when each contiguous element is small (instead of packing in such cases).  Further, the `ch4:ofi` layer constructs one IOV element at a time, making this problem even worse.  This is being handled in https://github.com/pmodels/mpich/pull/4475, although that PR is only looking at the `ch4:ofi` layer and not the AM layer.

**Comment by pavanbalaji** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by pavanbalaji** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by pavanbalaji** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by pavanbalaji** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by pavanbalaji** [comment]
test:mpich/ch4/ucx

**Comment by pavanbalaji** [comment]
test:mpich/ch4/ucx

**Comment by pavanbalaji** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by pavanbalaji** [comment]
test:mpich/ch4/ofi

**Comment by pavanbalaji** [comment]
test:mpich/ch4/ofi

**Comment by pavanbalaji** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by pavanbalaji** [comment]
With the `--with-datatype-engine=auto` configure option and the `ch4/ofi: Remove overly complex noncontiguous put/get` patch cherry-picked from https://github.com/pmodels/mpich/pull/4475, the status of testing is as follows:

1. `ch3/tcp` and `ch3/sock` passes all tests.  These use dataloop.
2. `ch4/ucx` passes all tests.
3. `ch4/ofi` passes all tests in the `default`, `am-only`, `strict`, `no-inline` and `external` configurations.

4. `ch4/ofi` in the `debug` configuration fails the following tests:

```
./rma/getfence1 2 -type=MPI_INT:4+MPI_DOUBLE:8 -count=512 -seed=136 -testsize=16
./rma/getfence1 2 -type=MPI_INT:4+MPI_DOUBLE:8 -count=1018 -seed=138 -testsize=16
./rma/getfence1 2 -type=MPI_INT -count=65530 -seed=139 -testsize=16
./rma/getfence1 2 -type=MPI_INT -count=262144 -seed=140 -testsize=16
./rma/getfence1 2 -type=MPI_INT:4+MPI_DOUBLE:8 -count=65530 -seed=141 -testsize=16
./rma/getfence1 2 -type=MPI_INT:4+MPI_DOUBLE:8 -count=262144 -seed=142 -testsize=16
./rma/putpscw1 4 -type=MPI_INT:4+MPI_DOUBLE:8 -count=262144 -seed=278 -testsize=16
```

... with these errors:

```
not ok  - ./rma/getfence1 2
  ---
  Directory: ./rma
  File: getfence1
  Num-procs: 2
  Timeout: 180
  Date: "Wed May  6 18:25:04 2020"
  ...
## Test output (expected 'No Errors'):
##  No Errors
## [1] 8192 at [0x888ae0], ./src/mpid/ch4/generic/am/mpidig_am_msg.h[266]
## [1] 4096 at [0x894e70], ./src/mpid/ch4/generic/am/mpidig_am_msg.h[266]
## [1] 2048 at [0x7b7950], ./src/mpid/ch4/generic/am/mpidig_am_msg.h[266]
## [1] 4096 at [0x898da0], ./src/mpid/ch4/generic/am/mpidig_am_msg.h[266]
## [1] 64 at [0x7b75b0], ./src/mpid/ch4/generic/am/mpidig_am_msg.h[266]
## [1] 4096 at [0x897cf0], ./src/mpid/ch4/generic/am/mpidig_am_msg.h[266]
## [1] 4096 at [0x892c30], ./src/mpid/ch4/generic/am/mpidig_am_msg.h[266]
## [1] 32 at [0x8280f0], ./src/mpid/ch4/generic/am/mpidig_am_msg.h[266]
## [1] 8192 at [0x886a30], ./src/mpid/ch4/generic/am/mpidig_am_msg.h[266]
## [1] 32 at [0x7b71f0], ./src/mpid/ch4/generic/am/mpidig_am_msg.h[266]
## [1] 8192 at [0x89a9a0], ./src/mpid/ch4/generic/am/mpidig_am_msg.h[266]
## [1] 8192 at [0x826040], ./src/mpid/ch4/generic/am/mpidig_am_msg.h[266]
## [1] 32 at [0x81b600], ./src/mpid/ch4/generic/am/mpidig_am_msg.h[266]
## [1] 8192 at [0x823f90], ./src/mpid/ch4/generic/am/mpidig_am_msg.h[266]
## [0] 4096 at [0xbc4770], ./src/mpid/ch4/generic/am/mpidig_am_msg.h[266]
## [0] 4096 at [0xbc36c0], ./src/mpid/ch4/generic/am/mpidig_am_msg.h[266]
## [0] 4096 at [0xc42420], ./src/mpid/ch4/generic/am/mpidig_am_msg.h[266]
## [0] 32 at [0xbcb140], ./src/mpid/ch4/generic/am/mpidig_am_msg.h[266]
## [0] 4096 at [0xbc2610], ./src/mpid/ch4/generic/am/mpidig_am_msg.h[266]
## [0] 4096 at [0xc41370], ./src/mpid/ch4/generic/am/mpidig_am_msg.h[266]
## [0] 64 at [0xbcadc0], ./src/mpid/ch4/generic/am/mpidig_am_msg.h[266]
## [0] 8192 at [0xc3b2b0], ./src/mpid/ch4/generic/am/mpidig_am_msg.h[266]
## [0] 8192 at [0xc39200], ./src/mpid/ch4/generic/am/mpidig_am_msg.h[266]
## [0] 32 at [0xbcaf80], ./src/mpid/ch4/generic/am/mpidig_am_msg.h[266]
## [0] 8192 at [0xc36c40], ./src/mpid/ch4/generic/am/mpidig_am_msg.h[266]
## [0] 4096 at [0xbc7810], ./src/mpid/ch4/generic/am/mpidig_am_msg.h[266]
## [0] 4096 at [0xbc9680], ./src/mpid/ch4/generic/am/mpidig_am_msg.h[266]
## [0] 4096 at [0xc43e30], ./src/mpid/ch4/generic/am/mpidig_am_msg.h[266]
## [0] 64 at [0xb5afe0], ./src/mpid/ch4/generic/am/mpidig_am_msg.h[266]
## [0] 64 at [0xbba1e0], ./src/mpid/ch4/generic/am/mpidig_am_msg.h[266]
## [0] 4096 at [0xbc6620], ./src/mpid/ch4/generic/am/mpidig_am_msg.h[266]
## [0] 32 at [0xb5b490], ./src/mpid/ch4/generic/am/mpidig_am_msg.h[266]
## [0] 4096 at [0xb59e40], ./src/mpid/ch4/generic/am/mpidig_am_msg.h[266]
```

These errors come from leaks originating from the cherry-picked commit from https://github.com/pmodels/mpich/pull/4475, which will be removed in the final merge.

4. `ch4/ofi` in the `debug` configuration times out in the following tests:

```
./rma/transpose5 2
./rma/derived_acc_flush_local 3
./rma/accfence1 4 -type=MPI_INT -count=65530 -seed=103 -testsize=16
./rma/accpscw1 4 -type=MPI_INT -count=65530 -seed=115 -testsize=16
./rma/lock_contention_dt 4 -type=MPI_INT -count=65530 -seed=153 -testsize=16
./rma/lock_dt 2 -type=MPI_INT -count=65530 -seed=162 -testsize=16
./rma/lock_dt_flush 2 -type=MPI_INT -count=65530 -seed=174 -testsize=16
./rma/lock_dt_flushlocal 2 -type=MPI_INT -count=65530 -seed=186 -testsize=16
./rma/lockall_dt 4 -type=MPI_INT -count=65530 -seed=201 -testsize=16
./rma/lockall_dt_flush 4 -type=MPI_INT -count=65530 -seed=215 -testsize=16
./rma/lockall_dt_flushall 4 -type=MPI_INT -count=65530 -seed=227 -testsize=16
./rma/lockall_dt_flushlocal 4 -type=MPI_INT -count=65530 -seed=237 -testsize=16
./rma/lockall_dt_flushlocalall 4 -type=MPI_INT -count=65530 -seed=249 -testsize=16
```

These timeouts are because of the performance bugs in `ch4/ofi` and `ch4/am` codes, which are being fixed in https://github.com/pmodels/mpich/pull/4475 and https://github.com/pmodels/mpich/pull/4498.

**Comment by pavanbalaji** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Pull request merged**